tests/allocator_api.rs

```diff
@@ -114,3 +114,31 @@ fn allocator_grow_zeroed() {
     assert!(bump.allocated_bytes() <= allocated * 2);
     assert_eq!(unsafe { p.as_ref() }, [42, 42, 42, 42, 0, 0, 0, 0]);
 }
+
+#[test]
+fn allocator_grow_layout_change() {
+    let b = AllocatorDebug::new(Bump::with_capacity(1024));
+
+    let layout_align4 = Layout::from_size_align(256, 4).unwrap();
+    let layout_align16 = Layout::from_size_align(1024, 16).unwrap();
+
+    // Allocate a chunk of memory and attempt to grow it while increasing
+    // alignment requirements.
+    let p4 = b.allocate(layout_align4).unwrap().as_non_null_ptr();
+    let _p16 = unsafe { b.grow(p4, layout_align4, layout_align16).unwrap() };
+}
+
+#[test]
+fn allocator_shrink_layout_change() {
+    let b = AllocatorDebug::new(Bump::with_capacity(1024));
+
+    let layout_align4 = Layout::from_size_align(1024, 4).unwrap();
+    let layout_align16 = Layout::from_size_align(256, 16).unwrap();
+
+    // Allocate a chunk of memory and attempt to shrink it while increasing
+    // alignment requirements.
+    let p4 = b.allocate(layout_align4).unwrap().as_non_null_ptr();
+    let p16_res = unsafe { b.shrink(p4, layout_align4, layout_align16) };
+
+    assert_eq!(p16_res, Err(AllocError));
+}
```

Comment:
Regarding being unsure whether the tests exercise the fallback-to-new-alloc paths, I think we could have much more confidence in exercising this code if we wrote targeted quickcheck tests instead of unit tests for these APIs.

I'm imagining that we would do something like:

```rust
quickcheck! {
    fn allocator_grow_layout_change(layouts: Vec<(usize, usize)>) -> () {
        let mut layouts = layouts.into_iter().map(|(size, align)| {
            let align = std::cmp::min(align, 64);
            let align = round_up_to_pow2(align);
            let size = std::cmp::min(size, 1024);
            let size = std::cmp::max(size, align);
            Layout::from_size_align(size, align).unwrap()
        }).collect::<Vec<_>>();
        layouts.sort_by_key(|l| (l.size(), l.align()));
        // TODO: allocate the first layout in the bump and then grow
        // to each of the next layouts
    }
}
```

where `round_up_to_pow2` is basically this helper but without checked arithmetic (safe because we are clamping the input just above): https://cs.github.com/bytecodealliance/wasmtime/blob/b438617e1238bec01ebc804c0087e4849df0dc87/crates/runtime/src/externref.rs?q=round_up#L291

Exercising shrinking would be the exact same except that we would invert the comparator so that the layouts were ordered form largest to smallest.

Does that make sense?

Reply:
Yes, this makes sense. Do we want to assert something in these quickchecks, or is it enough that they don't panic for now?

Reply:
I think we can start with just asserting that the grown/shrunk pointer has the requested alignment.